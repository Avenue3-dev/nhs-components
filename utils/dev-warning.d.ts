/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
