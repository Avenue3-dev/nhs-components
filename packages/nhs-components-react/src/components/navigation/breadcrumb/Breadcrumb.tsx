import React, { HTMLProps, ReactNode } from 'react'
import { AsElementLink } from '@/types/link-like'
import clsx from 'clsx'

type BreadcrumbItem = React.FC<AsElementLink<HTMLAnchorElement>>
type BreadcrumbBackItem = React.FC<AsElementLink<HTMLAnchorElement>>

/**
 * Use breadcrumbs to help users understand where they are in the website.
 *
 * See more information regarding this component [here](https://service-manual.nhs.uk/design-system/components/breadcrumbs).
 */
const Item: BreadcrumbItem = ({
  className,
  children,
  asElement: Component = 'a',
  ...rest
}): JSX.Element => {
  return (
    <li className="nhsuk-breadcrumb__item">
      <Component
        className={clsx('nhsuk-breadcrumb__link', className)}
        {...rest}
      >
        {children}
      </Component>
    </li>
  )
}

const BackItem: BreadcrumbBackItem = ({
  className,
  children,
  asElement: Component = 'a',
  ...rest
}): JSX.Element => {
  return (
    <p className="nhsuk-breadcrumb__back">
      <Component
        className={clsx('nhsuk-breadcrumb__backlink', className)}
        {...rest}
      >
        <span className="nhsuk-u-visually-hidden">Back to &nbsp;</span>
        {children}
      </Component>
    </p>
  )
}

type Breadcrumb = {
  Item: BreadcrumbItem
  Back: BreadcrumbBackItem
} & React.FC<HTMLProps<HTMLDivElement>>

type Split = {
  Items: ReactNode[]
  Back?: ReactNode
  Others: ReactNode[]
}

const Breadcrumb: Breadcrumb = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  const { Items, Back, Others } = React.Children.toArray(
    children,
  ).reduce<Split>(
    (prev, child) => {
      if (child && typeof child === 'object' && 'type' in child) {
        if (child.type === Item) {
          prev.Items.push(child)
        } else if (child.type === BackItem) {
          prev.Back = child
        } else {
          prev.Others.push(child)
        }
      } else {
        prev.Others.push(child)
      }

      return prev
    },
    { Items: [], Back: undefined, Others: [] },
  )

  return (
    <nav className={clsx('nhsuk-breadcrumb', className)} {...rest}>
      <div className="nhsuk-width-container">
        <ol className="nhsuk-breadcrumb__list">{Items}</ol>
        {Back}
        {Others}
      </div>
    </nav>
  )
}

Breadcrumb.Back = BackItem
Breadcrumb.Item = Item

Breadcrumb.defaultProps = {
  'aria-label': 'Breadcrumb',
}

export default Breadcrumb
