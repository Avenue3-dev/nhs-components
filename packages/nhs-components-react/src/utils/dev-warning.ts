const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
