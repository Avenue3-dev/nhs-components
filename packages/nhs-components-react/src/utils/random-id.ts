export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
