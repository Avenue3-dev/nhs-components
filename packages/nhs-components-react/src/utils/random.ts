export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
