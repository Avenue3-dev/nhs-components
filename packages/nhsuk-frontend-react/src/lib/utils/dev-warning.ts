const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
