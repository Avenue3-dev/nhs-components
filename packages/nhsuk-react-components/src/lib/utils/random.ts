export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
