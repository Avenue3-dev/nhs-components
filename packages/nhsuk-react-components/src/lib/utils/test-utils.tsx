/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
