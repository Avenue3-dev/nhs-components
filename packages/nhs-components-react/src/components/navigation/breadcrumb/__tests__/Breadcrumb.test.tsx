import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import { Link } from 'react-router-dom'
import Breadcrumb from '..'

describe('Breadcrumb', () => {
  test('should match the snapshot', () => {
    render(
      <Breadcrumb data-testid="breadcrumb">
        <Breadcrumb.Item href="/url/1">Item 1</Breadcrumb.Item>
        <Breadcrumb.Item href="/url/2">Item 1</Breadcrumb.Item>
        <Breadcrumb.Back href="/url/1">Item 1</Breadcrumb.Back>
      </Breadcrumb>,
    )

    expect(screen.getByTestId('breadcrumb')).toMatchSnapshot()
  })

  test('should match the snapshot - other elements', () => {
    render(
      <Breadcrumb data-testid="breadcrumb">
        <Breadcrumb.Item href="/url/1">Item 1</Breadcrumb.Item>
        <Breadcrumb.Item href="/url/2">Item 1</Breadcrumb.Item>
        <Breadcrumb.Back href="/url/1">Item 1</Breadcrumb.Back>
        <>Item</>
        item
      </Breadcrumb>,
    )

    expect(screen.getByTestId('breadcrumb')).toMatchSnapshot()
  })

  test('accepts react-router-dom Link element', () => {
    render(
      <Breadcrumb data-testid="breadcrumb">
        <Breadcrumb.Item asElement={Link} to="/url/1">
          Item 1
        </Breadcrumb.Item>
        <Breadcrumb.Item asElement={Link} to="/url/2">
          Item 1
        </Breadcrumb.Item>
        <Breadcrumb.Back asElement={Link} to="/url/1">
          Item 1
        </Breadcrumb.Back>
      </Breadcrumb>,
    )

    expect(screen.getByTestId('breadcrumb')).toMatchSnapshot()
  })
})
