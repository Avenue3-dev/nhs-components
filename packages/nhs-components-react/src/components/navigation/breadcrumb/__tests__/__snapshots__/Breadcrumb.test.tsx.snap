// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb > accepts react-router-dom Link element 1`] = `
<nav
  aria-label="Breadcrumb"
  class="nhsuk-breadcrumb"
  data-testid="breadcrumb"
>
  <div
    class="nhsuk-width-container"
  >
    <ol
      class="nhsuk-breadcrumb__list"
    >
      <li
        class="nhsuk-breadcrumb__item"
      >
        <a
          class="nhsuk-breadcrumb__link"
          href="/url/1"
        >
          Item 1
        </a>
      </li>
      <li
        class="nhsuk-breadcrumb__item"
      >
        <a
          class="nhsuk-breadcrumb__link"
          href="/url/2"
        >
          Item 1
        </a>
      </li>
    </ol>
    <p
      class="nhsuk-breadcrumb__back"
    >
      <a
        class="nhsuk-breadcrumb__backlink"
        href="/url/1"
      >
        <span
          class="nhsuk-u-visually-hidden"
        >
          Back to  
        </span>
        Item 1
      </a>
    </p>
  </div>
</nav>
`;

exports[`Breadcrumb > should match snapshot 1`] = `
<nav
  aria-label="Breadcrumb"
  class="nhsuk-breadcrumb"
  data-testid="breadcrumb"
>
  <div
    class="nhsuk-width-container"
  >
    <ol
      class="nhsuk-breadcrumb__list"
    >
      <li
        class="nhsuk-breadcrumb__item"
      >
        <a
          class="nhsuk-breadcrumb__link"
          href="/url/1"
        >
          Item 1
        </a>
      </li>
      <li
        class="nhsuk-breadcrumb__item"
      >
        <a
          class="nhsuk-breadcrumb__link"
          href="/url/2"
        >
          Item 1
        </a>
      </li>
    </ol>
    <p
      class="nhsuk-breadcrumb__back"
    >
      <a
        class="nhsuk-breadcrumb__backlink"
        href="/url/1"
      >
        <span
          class="nhsuk-u-visually-hidden"
        >
          Back to  
        </span>
        Item 1
      </a>
    </p>
  </div>
</nav>
`;
